### ISSUE TITLE
Unexpected zero page static stack behavior

### ISSUE THREAD

**Issue by atn34** [llvm-mos/llvm-mos#210]
```
$ mos-dodo-clang++ --version
clang version 15.0.0 (https://www.github.com/llvm-mos/llvm-mos.git 17116f6907fb1ccbf0c5a6e43d8bb6be088bc48f)
...

$ cat test.cpp 
#include <api.h>

unsigned char get_buttons() {
  unsigned char buttons = READ_BUTTONS();
  return buttons;
}

int main() { get_buttons(); }

$ mos-dodo-clang++ test.cpp && llvm-objdump -d --print-imm-hex a.out.elf | grep get_buttons -A 3
0000591d <_Z11get_buttonsv>:
    591d: 20 18 59      jsr     $5918
    5920: 8d 00 02      sta     $200
    5923: 60            rts
```

The local variable "buttons" in "get_buttons" is placed in the static stack, but I expected it to be placed in the zero page static stack.

Interestingly, if we initialize buttons with `0` instead `READ_BUTTONS` we do get the zero page static stack, but it appears to get assigned to address `$0`, which is not available for general use in dodo.

```
$ mos-dodo-clang++ test.cpp && llvm-objdump -d --print-imm-hex a.out.elf | grep get_buttons -A 3
00005918 <_Z11get_buttonsv>:
    5918: a2 00         ldx     #$0
    591a: 86 00         stx     $0
    591c: 60            rts
```

**Comment by mysterymath** [comment]
Did a quick look, didn't take very long to track down what's happening here.

The root of the problem is that the compiler doesn't know anything about what goes on in READ_BUTTONS. For all it knows, READ_BUTTONS could call `main()`, which could in turn call `get_buttons()`, which would make the whole thing recursive.

At least this would be the story in C. In C++, the language specification specifically makes it undefined behavior for main to recurse. So, main gets marked as non-recursive, which marks get_buttons as non-recursive. So static stack can be allocated.

Trouble is, I took a few shortcuts in how zero page allocation works that require a clearer view of what can call what than exists in this case. So it still bails, even though zero page stack could actually be allocated in this case.

Accordingly, there's a few solutions here.
First, the routines in api.h that cannot call back to anything in the main LTO unit (probably they don't have any calls at all) should be marked as `__attribute__((leaf))`. This greatly simplifies the ZP and stack analysis, which should allow this pattern to work in e.g. C.

Second,  I should go through and fix the cheats so that zero page allocation can work, even in corner cases like this. I was rolling the dice on whether an issue of this class would ever be found, and one was found five minutes after the feature was released, so that settles that.
